cpp/Esri/ArcGISRuntime/Toolkit/GenericListModel.cpp

```diff
@@ -0,0 +1,379 @@
+    [this](const QModelIndex& topLeft, const QModelIndex& bottomRight,
+       const QVector<int>& roles)
+  {
+    if (m_displayPropIndex < 0)
+      return;
+
+    if (!m_elementType)
+      return;
+
+    if (roles.contains(Qt::DisplayRole))
+      return;
+
+    const auto offset = m_elementType->propertyOffset();
+    if (roles.contains(m_displayPropIndex - offset + Qt::UserRole +1))
+      emit dataChanged(topLeft, bottomRight, QVector<int>() << Qt::DisplayRole << Qt::EditRole);
+  });
+}
+
+GenericListModel::~GenericListModel()
+{
+}
+
+int GenericListModel::rowCount(const QModelIndex& parent) const
+{
+  if (parent.isValid())
+    return 0;
+  else
+    return m_objects.size();
+}
+
+QVariant GenericListModel::data(const QModelIndex& index, int role) const
+{
+  if (!m_elementType)
+    return QVariant();
+  else if (!index.isValid())
+    return QVariant();
+
+  auto o = m_objects.at(index.row());
+  if (role == Qt::DisplayRole || role == Qt::EditRole)
+  {
+    const auto property = m_elementType->property(m_displayPropIndex);
+    return property.read(o);
+  }
+  else if (role == Qt::UserRole)
+  {
+    return QVariant::fromValue<QObject*>(o);
+  }
+  else if (role >= Qt::UserRole)
+  {
+    const auto propIndex = role - (Qt::UserRole + 1) + m_elementType->propertyOffset();
+    const auto property = m_elementType->property(propIndex);
+    return property.read(o);
+  }
+  else
+  {
+    return QVariant();
+  }
+}
+
+bool GenericListModel::setData(const QModelIndex& index, const QVariant& value, int role)
+{
+  if (!m_elementType)
+    return false;
+  else if (!index.isValid())
+    return false;
+
+  if (role == Qt::DisplayRole || role == Qt::EditRole)
+  {
+    auto o = m_objects.at(index.row());
+    const auto property = m_elementType->property(m_displayPropIndex);
+    return property.write(o, value);
+  }
+  else if (role == Qt::UserRole)
+  {
+    auto newObject = qvariant_cast<QObject*>(value);
+    if (newObject && m_elementType == newObject->metaObject())
+    {
+      m_objects[index.row()] = newObject;
+      emit dataChanged(index, index);
+      return true;
+    }
+    else
+    {
+      return false;
+    }
+  }
+  else if (role > Qt::UserRole)
+  {
+    auto o = m_objects.at(index.row());
+    const auto propIndex = role - (Qt::UserRole + 1) + m_elementType->propertyOffset();
+    const auto property = m_elementType->property(propIndex);
+    return property.write(o, value);
+  }
+  else
+  {
+    return false;
+  }
+}
+
+QHash<int, QByteArray> GenericListModel::roleNames() const
+{
+  if (!m_elementType)
+    return { };
+
+  QHash<int, QByteArray> output;
+  output.insert( Qt::UserRole, "modelData");
+
+  const int offset = m_elementType->propertyOffset();
+  for (int i = offset; i < m_elementType->propertyCount(); ++i)
+  {
+    auto property = m_elementType->property(i);
+    output.insert(i - offset + Qt::UserRole + 1, property.name());
+  }
+  return output;
+}
```

Comment:
There's a lot of meta object magic going on here. Can you give a quick overview of how this works and what it's doing?

Reply:
In the code? Yes good idea I'll add a comment.